src/wayland/session_lock/lock.rs

```diff
@@ -0,0 +1,142 @@
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            Request::GetLockSurface { id, surface, output } => {
+                // Assign surface a role and ensure it never had one before.
+                if compositor::give_role(&surface, LOCK_SURFACE_ROLE).is_err() {
+                    lock.post_error(Error::Role, "Surface already has a role.");
+                    return;
+                }
+
+                // Ensure output is not already locked.
+                let lock_state = state.lock_state();
+                if lock_state.locked_outputs.contains(&output) {
+                    lock.post_error(Error::DuplicateOutput, "Output is already locked.");
+                    return;
+                }
+                lock_state.locked_outputs.push(output.clone());
+
+                // Ensure surface has no existing buffers attached.
+                let has_buffer = compositor::with_states(&surface, |states| {
+                    let cached = &states.cached_state;
+                    let pending = cached.pending::<SurfaceAttributes>().buffer.is_some();
+                    let current = cached.current::<SurfaceAttributes>().buffer.is_some();
+                    pending || current
+                });
+                if has_buffer {
+                    lock.post_error(Error::AlreadyConstructed, "Surface has a buffer attached.");
+                    return;
+                }
+
+                let data = ExtLockSurfaceUserData {
+                    surface: surface.clone(),
+                };
+                let lock_surface = data_init.init(id, data);
+
+                // Initialize surface data.
+                compositor::with_states(&surface, |states| {
+                    let inserted = states.data_map.insert_if_missing_threadsafe(|| {
+                        Mutex::new(LockSurfaceAttributes::new(lock_surface.clone()))
+                    });
+
+                    if !inserted {
+                        let mut attributes = states
+                            .data_map
+                            .get::<Mutex<LockSurfaceAttributes>>()
+                            .unwrap()
+                            .lock()
+                            .unwrap();
+                        attributes.surface = lock_surface.clone();
+                    }
+                });
+
+                // Add pre-commit hook for updating surface state.
+                compositor::add_pre_commit_hook::<D, _>(&surface, |_state, _dh, surface| {
+                    compositor::with_states(surface, |states| {
+                        let attributes = states.data_map.get::<Mutex<LockSurfaceAttributes>>();
+                        let mut attributes = attributes.unwrap().lock().unwrap();
+
+                        let Some(state) = attributes.last_acked else {
+                            attributes.surface.post_error(
+                                ext_session_lock_surface_v1::Error::CommitBeforeFirstAck,
```

Comment:
The protocol also defines `null_buffer` and `dimensions_mismatch` protocol errors. It doesn't seem to mention them, but presumably we should have a check here that a null buffer is committed and it's dimensions match what was given to `configure`.

Reply:
Done.

For the NULL buffer I'm checking for `BufferAssignment::Removed`, which as far as I understand doesn't cover committing with no buffer in the first place, but I'm not sure how to easily check for that.

For dimensions I think this is right? I tested with scale 1 and 2 and dynamically resizing the output, seems to work.